typescript/src/entities/product.ts

```diff
@@ -139,4 +139,138 @@ export class Product extends Entity {
 
   @Product.property({arrayType: "User"})
   public suppliers?: Array<User>;
+
+  public duplicate = () => {
+    /* create a clone of this product on the backend, returning it. */
+    const resourceName = (this.constructor as typeof Product).resourceName;
+    const resource = `/${resourceName}/${String(this.id)}/copy/`;
+    const fetchOptions = {method: 'POST'};
+    return this.merchi.authenticatedFetch(resource, fetchOptions).
+      then((data: any) => {
+        const product = new this.merchi.Product();
+        product.fromJson(data);
+        return product;
+      });
+  }
+
+  public primaryImage = () => {
+    if (this.featureImage === undefined) {
+      throw new Error("featureImage is undefined, did you forget to embed it?");
+    }
+    if (this.images === undefined) {
+      throw new Error("images is undefined, did you forget to embed it?");
+    }
+    if (this.featureImage !== null) {
+      return this.featureImage;
+    }
+    if (this.images.length > 0) {
+      return this.images[0];
+    }
+    return null;
+  }
+
+  public currency = () => {
+    if (this.domain === undefined) {
+      throw new Error("domain is undefined, did you forget to embed it?");
+    }
+    return this.domain.defaultCurrency();
+  }
+
+  public hasGroupVariationFields = () => {
+    if (this.groupVariationFields === undefined) {
+      const err = "groupVariationFields is undefined, did you forget to embed" +
+        " it?";
+      throw new Error(err);
+    }
+    return this.groupVariationFields.length > 0;
+  }
+
+  public hasIndependentVariationFields = () => {
+    if (this.independentVariationFields === undefined) {
+      const err = "independentVariationFields is undefined, did you forget to" +
+        " embed it?";
+      throw new Error(err);
+    }
+    return this.independentVariationFields.length > 0;
+  }
+
+  public taxType = () => {
+    if (this.domain === undefined) {
```

Comment:
can wrap in a function checkDomainEmbeded()

Reply:
not until typescript supports "assertion signatures" (i.e., 3.7)